service/src/main/java/bio/terra/tanagra/db/ActivityLogDao.java

```diff
@@ -0,0 +1,240 @@
+  private static final RowMapper<Pair<String, ActivityLogResource>>
+      ACTIVITY_LOG_RESOURCE_ROW_MAPPER =
+          (rs, rowNum) ->
+              Pair.of(
+                  rs.getString("activity_log_id"),
+                  ActivityLogResource.builder()
+                      .type(ActivityLogResource.Type.valueOf(rs.getString("resource_type")))
+                      .studyId(rs.getString("study_id"))
+                      .cohortId(rs.getString("cohort_id"))
+                      .cohortRevisionId(rs.getString("cohort_revision_id"))
+                      .reviewId(rs.getString("review_id"))
+                      .studyDisplayName(rs.getString("study_display_name"))
+                      .cohortDisplayName(rs.getString("cohort_display_name"))
+                      .reviewDisplayName(rs.getString("review_display_name"))
+                      .build());
+
+  private final NamedParameterJdbcTemplate jdbcTemplate;
+
+  @Autowired
+  public ActivityLogDao(NamedParameterJdbcTemplate jdbcTemplate) {
+    this.jdbcTemplate = jdbcTemplate;
+  }
+
+  @WriteTransaction
+  public void createActivityLog(ActivityLog activityLog) {
+    String sql =
+        "INSERT INTO activity_log (id, user_email, version_git_tag, version_git_hash, version_build, activity_type, export_model) "
+            + "VALUES (:id, :user_email, :version_git_tag, :version_git_hash, :version_build, :activity_type, :export_model)";
+    LOGGER.debug("CREATE activity log: {}", sql);
+    MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue("id", activityLog.getId())
+            .addValue("user_email", activityLog.getUserEmail())
+            .addValue("version_git_tag", activityLog.getVersionGitTag())
+            .addValue("version_git_hash", activityLog.getVersionGitHash())
+            .addValue("version_build", activityLog.getVersionBuild())
+            .addValue("activity_type", activityLog.getType().name())
+            .addValue("export_model", activityLog.getExportModel());
+    int rowsAffected = jdbcTemplate.update(sql, params);
+    LOGGER.debug("CREATE activity log rowsAffected = {}", rowsAffected);
+
+    if (!activityLog.getResources().isEmpty()) {
+      sql =
+          "INSERT INTO activity_log_resource (activity_log_id, stable_index, resource_type, study_id, cohort_id, cohort_revision_id, review_id) "
+              + "VALUES (:activity_log_id, :stable_index, :resource_type, :study_id, :cohort_id, :cohort_revision_id, :review_id)";
+      LOGGER.debug("CREATE activity log resource: {}", sql);
+      List<MapSqlParameterSource> resourceParamSets = new ArrayList<>();
+      for (int i = 0; i < activityLog.getResources().size(); i++) {
+        ActivityLogResource resource = activityLog.getResources().get(i);
+        resourceParamSets.add(
+            new MapSqlParameterSource()
+                .addValue("activity_log_id", activityLog.getId())
+                .addValue("stable_index", i)
```

Comment:
Is a stable index actually needed here? Won't the default ordering be by date/time anyway?

Reply:
This column is for ordering the list of resources for each activity log entry. You're right that the default ordering for the log entries themselves will always be by the `logged` timestamp, so no need for a stable index on that table.